### DISCUSSION TITLE
Bug: RangeError when using renderer.plugins.extract.image

### DISCUSSION THREAD

**Discussion by bildungsroman** [pixijs/pixijs#9536] [Q&A] [answered]
### Current Behavior

I'm trying to export the entire canvas as a PNG using PIXI.js in a React app that also uses `react-pixi`. I'm on version 6.5.9 and trying the following:

```javascript
// MyComponent.tsx

<button onClick={exportImage} />

// utils.ts
import { useApp } from "@pixi/react";

export const exportImage = () => {
  const app = useApp(); // PIXI.Application
  if (app) {
    app.renderer.plugins.extract.image(app.stage, "image/png", 1.0);
    const createEl = document.createElement("a");
    createEl.href = image.src;
    createEl.download = "my-canvas";
    createEl.click();
    createEl.remove();
  }
}
```

However, no PNG image is created and instead I'm getting this error in the console:

```bash
Uncaught RangeError: Invalid typed array length: 9012016008004
    at new Uint8Array (<anonymous>)
    at Extract2._rawPixels (Extract.ts:220:24)
    at Extract2.canvas (Extract.ts:107:55)
    at Extract2.base64 (Extract.ts:95:21)
    at Extract2.image (Extract.ts:79:26)
```

### Expected Behavior

I would expect a PNG to be saved to disk with no errors.

### Steps to Reproduce

In a PIXI app, try `app.renderer.plugins.extract.image(app.stage, "image/png", 1.0);`

### Environment

- **`pixi.js` version**: *6.5.9*
- **Browser & Version**: *Chrome 114*
- **OS & Version**: *M1, macOS 13.4*
- **Running Example**: *n/a*

**Comment by bigtimebuddy** [comment]
The basic functionality worked fine for me:
https://jsfiddle.net/zum4Lgro/

I'm guessing this is because your canvas is VERY big. The Uint8Array is the width * height * 4. Check your Application setting for the width/height and resolution.

If that fails, could you make a minimal reproduction that recreate your error?

**Comment by SignDawn** [comment]
这没有办法，当 canvas 过大时，导出图片就可能出现问题，这是浏览器的问题。
或许该想办法减小 canvas 的宽高。
demo 可以查看下面这个，当 side 宽度达到 20000 时就会出现问题，若无法出现，可以尝试设置更大的值。
https://www.pixiplayground.com/#/edit/cx2K0N7AnmGXYt55fX59j

**Comment by SignDawn** [comment]
@bigtimebuddy this demo 
https://www.pixiplayground.com/#/edit/cx2K0N7AnmGXYt55fX59j

**Comment by bigtimebuddy** [comment]
The example is trying to extract a 2M x 2M pixel image in that example which, on some devices, is not possible. If you just want to export only what's in the main view, instead of this:

```js
app.renderer.extract.base64(app.stage);
```
do this instead:
```js
app.renderer.extract.base64();
```

This will only export the main renderer's contents and not the entire stage Container.

If you need to extract a very large displayobject, the best way to do this and to not melt your user's computer, is to export it as tiles and then use another tool like imagemagick to stitch them back together.

**Comment by bildungsroman** [comment]
Thanks for your answer, @bigtimebuddy. I tried the following:

```javascript
        const image = app.renderer.plugins.extract.base64();
        const createEl = document.createElement("a");
        createEl.href = image.src;
        createEl.download = "my-canvas";
        createEl.click();
        createEl.remove();
```

and I get an `HTML` element rather than a PNG. If I do `createEl.href = image.toDataURL();`, I get `image.toDataURL is not a function`.

Another thing I have tried is `const image = app.renderer.plugins.extract.image();`. This works, but the PNG I get is blank. It's doubly annoying because I can just right-click on the canvas in the browser, select "Save Image As...", and I get exactly what I want as an image.

**Comment by bildungsroman** [comment]
@SignDawn You're right, the stage is too large. When I log `app.stage.height` I get `1501001` 😬 Is there a way to reduce the stage size to just the visible canvas?

**Comment by bildungsroman** [comment]
I'm able to do this without the range error, but I'm still getting a blank image when the PNG is saved. What am I doing wrong here?

```javascript
        const canvas = app.renderer.plugins.extract.canvas(app.stage, {
          width: 3840,
          height: 2160,
        });
        const canvasUrl = canvas.toDataURL();
        const createEl = document.createElement("a");
        createEl.href = canvasUrl;
        createEl.download = "my-canvas";
        createEl.click();
        createEl.remove();
```

I am getting the following warnings in the console: 

```bash
1 [.WebGL-0x1040497f500] GL_INVALID_VALUE: Desired resource size is greater than max renderbuffer size.
100[.WebGL-0x1040497f500] GL_INVALID_FRAMEBUFFER_OPERATION: Draw framebuffer is incomplete
1 [.WebGL-0x1040497f500] GL_INVALID_FRAMEBUFFER_OPERATION: Framebuffer is incomplete: Attachment has zero size.
```

**Comment by bigtimebuddy** [comment]
> What am I doing wrong here?

Don't pass `app.stage` to `canvas()`, `image()` etc, this will use only what's in the view.

**Comment by bildungsroman** [comment]
@bigtimebuddy Ok, when I do it without passing `app.stage`, it works but I get a blank image:

```javascript
        const canvas = app.renderer.plugins.extract.canvas();
        const canvasUrl = canvas.toDataURL();
        const createEl = document.createElement("a");
        createEl.href = canvasUrl;
        createEl.download = "my-canvas";
        createEl.click();
        createEl.remove();
```

Also, [this answer](https://github.com/pixijs/pixijs/issues/3880#issuecomment-289225134) made me think that passing `stage` is required.

I don't have my own playground set up, I'm I'm able to reproduce the same things I'm seeing in this old example: https://stackblitz.com/edit/angular-xmu6se-pixijs-dil7ja?file=src%2Fapp%2Fapp.component.ts

When I click `Save PNG` I get a blank file. However, when I right-click and do `Save As Image...` I get a proper PNG of the canvas.

**Comment by bildungsroman** [comment, marked as answer]
After much struggle (and help from a colleague), I got it to work by passing in the viewport and limiting it to `app.screen`. I'll post my code to help future searchers who may run into this:

```javascript
// MyComponent.tsx

<button onClick={exportImage} />

// utils.ts
import { useApp } from "@pixi/react";

export const exportImage = () => {
  const app = useApp(); // PIXI.Application
  const viewport = new PixiViewport(options);

  const imageThis = app.renderer.generateTexture(viewport, {
    region: app.screen,
    resolution: 2,
  });
  const image: HTMLImageElement = app.renderer.plugins.extract.image(
    imageThis,
    "image/png",
    1.0
  );
  const createEl = document.createElement("a");
  createEl.href = image.src;
  createEl.download = "my-canvas";
  createEl.click();
  createEl.remove();
  imageThis.destroy();
};
```
